src/rgw/rgw_d3n_datacache.h

```diff
@@ -0,0 +1,366 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+#ifndef CEPH_RGWD3NDATACACHE_H
+#define CEPH_RGWD3NDATACACHE_H
+
+#include "rgw_rados.h"
+#include <curl/curl.h>
+
+#include "rgw_common.h"
+
+#include <unistd.h>
+#include <signal.h>
+#include "include/Context.h"
+#include "include/lru.h"
+#include "rgw_threadpool.h"
+#include "rgw_d3n_cacherequest.h"
+
+
+/*D3nDataCache*/
+struct D3nDataCache;
+class D3nL2CacheThreadPool;
+class D3nHttpL2Request;
+
+struct D3nChunkDataInfo : public LRUObject {
+	CephContext *cct;
+	uint64_t size;
+	time_t access_time;
+	string address;
+	string oid;
+	bool complete;
+	struct D3nChunkDataInfo* lru_prev;
+	struct D3nChunkDataInfo* lru_next;
+
+	D3nChunkDataInfo(): size(0) {}
+
+	void set_ctx(CephContext *_cct) {
+		cct = _cct;
+	}
+
+	void dump(Formatter *f) const;
+	static void generate_test_instances(list<D3nChunkDataInfo*>& o);
+};
+
+struct D3nCacheAioWriteRequest{
+	string oid;
+	void *data;
+	int fd;
+	struct aiocb *cb;
+	D3nDataCache *priv_data;
+	CephContext *cct;
+
+	D3nCacheAioWriteRequest(CephContext *_cct) : cct(_cct) {}
+	int create_io(bufferlist& bl, unsigned int len, string oid, string cache_location);
+
+	void release() {
+		::close(fd);
+		cb->aio_buf = nullptr;
+		free(data);
+		data = nullptr;
+		free(cb);
+		free(this);
+	}
+};
+
+struct D3nDataCache {
+
+private:
+  std::map<string, D3nChunkDataInfo*> cache_map;
+  std::list<string> outstanding_write_list;
```

Comment:
I wonder if this could be improved (perhaps in a later refactoring) with boost::intrusive.  An intrusive cache map is generally going to be more efficient than a std::map of same, and there might be a way to do something with the outstanding_write_list too--which I don't fully grok tbh, but looks suspicious (list of names that may be keys into the cache_map?)

Reply:
in https://github.com/ceph/ceph/pull/36266/files#r537855197 i propose a design that doesn't need this list

Reply:
changing to `std::set` for now